src/main/scala/io/iohk/ethereum/network/AuthHandshaker.scala

```diff
@@ -0,0 +1,204 @@
+import org.spongycastle.crypto.digests.{KeccakDigest, SHA256Digest}
+import org.spongycastle.crypto.params.{ECPrivateKeyParameters, ECPublicKeyParameters}
+import org.spongycastle.crypto.signers.{HMacDSAKCalculator, ECDSASigner}
+import org.spongycastle.math.ec.ECPoint
+import scorex.core.network
+import scorex.core.network._
+
+class Secrets(
+    val aes: Array[Byte],
+    val mac: Array[Byte],
+    val token: Array[Byte],
+    val egressMac: KeccakDigest,
+    val ingressMac: KeccakDigest)
+  extends scorex.core.network.Secrets
+
+object AuthHandshaker {
+
+  val NonceSize = 32
+  val MacSize = 256
+  val SecretSize = 32
+
+  def apply(nodeKey: AsymmetricCipherKeyPair): AuthHandshaker = {
+    val nonce = new Array[Byte](NonceSize)
+    new SecureRandom().nextBytes(nonce)
+    AuthHandshaker(nodeKey, ByteString(nonce))
+  }
+
+}
+
+case class AuthHandshaker(
+    nodeKey: AsymmetricCipherKeyPair,
+    nonce: ByteString,
+    ephemeralKey: AsymmetricCipherKeyPair = generateKeyPair(),
+    isInitiator: Boolean = false,
+    initiateMessageOpt: Option[AuthInitiateMessage] = None,
+    initiatePacketOpt: Option[ByteString] = None,
+    responseMessageOpt: Option[AuthResponseMessage] = None,
+    responsePacketOpt: Option[ByteString] = None)
+  extends scorex.core.network.AuthHandshaker {
+
+  import AuthHandshaker._
+
+  override def initiate(uri: URI): (ByteString, network.AuthHandshaker) = {
+    val remotePubKey = publicKeyFromNodeId(uri.getUserInfo)
+    val message = createAuthInitiateMessage(remotePubKey)
+    val encryptedPacket = ByteString(ECIESCoder.encrypt(remotePubKey, message.encode().toArray, None))
+
+    (encryptedPacket, copy(isInitiator = true, initiateMessageOpt = Some(message), initiatePacketOpt = Some(encryptedPacket)))
+  }
+
+  override def handleResponseMessage(data: ByteString): AuthHandshakeResult = {
+    val plaintext = ECIESCoder.decrypt(nodeKey.getPrivate.asInstanceOf[ECPrivateKeyParameters].getD, data.toArray)
+    val message = AuthResponseMessage.decode(plaintext)
+
+    copy(
+      responseMessageOpt = Some(message),
+      responsePacketOpt = Some(data)).finalizeHandshake()
+  }
+
+  override def handleInitialMessage(data: ByteString): (ByteString, AuthHandshakeResult) = {
+    val plaintext = ECIESCoder.decrypt(nodeKey.getPrivate.asInstanceOf[ECPrivateKeyParameters].getD, data.toArray)
+    val message = AuthInitiateMessage.decode(plaintext)
+
+    val response = createAuthResponseMessage()
+    val encryptedPacket = ByteString(ECIESCoder.encrypt(message.publicKey, response.encode().toArray, None))
+
+    val handshakeResult = copy(
+      initiateMessageOpt = Some(message),
+      initiatePacketOpt = Some(data),
+      responseMessageOpt = Some(response),
+      responsePacketOpt = Some(encryptedPacket)).finalizeHandshake()
+
+    (encryptedPacket, handshakeResult)
+  }
+
+  private def createAuthInitiateMessage(remotePubKey: ECPoint) = {
+    // TODO: handle the case when the peer is known
```

Comment:
Yes. The logic is slightly different when the peer is known (session tokens must be stored somewhere).